src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -3455,6 +3487,110 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
         }
     }
 }
+//----------------------------------------------------------------------------------------------
+// Lowering::LowerHWIntrinsicCndSel: Lowers a Sve ConditionalSelect call
+//
+//  Arguments:
+//     node - The hardware intrinsic node of the form
+//            ConditionalSelect(mask, trueValue, falseValue)
+//
+GenTree* Lowering::LowerHWIntrinsicCndSel(GenTreeHWIntrinsic* cndSelNode)
+{
+    assert(cndSelNode->OperIsHWIntrinsic(NI_Sve_ConditionalSelect));
+
+    GenTree* op1 = cndSelNode->Op(1);
+    GenTree* op2 = cndSelNode->Op(2);
+    GenTree* op3 = cndSelNode->Op(3);
+
+    if (op1->OperIsHWIntrinsic())
+    {
+        GenTreeHWIntrinsic* predNode = op1->AsHWIntrinsic();
+        if (predNode->IsAllFalse())
+        {
+            // Transform:
+            //
+            // CndSel(AllFalse, op2, op3) to
+            // op3
+            op1->SetUnusedValue();
+            op2->SetUnusedValue();
+            cndSelNode->SetUnusedValue();
+
+            return op3;
+        }
+
+        if (predNode->IsAllTrue())
+        {
+            if (op2->isEmbeddedMaskingCompatibleHWIntrinsic())
+            {
+                // Transform:
+                //
+                // CndSel(AllTrue, embeddedMask(op2), op3) to
+                // CndSel(AllTrue, embeddedMask(op2), Zero)
+                //
+                // Zero will get contained during ConditionalSelect
+                // containment
+
+                if (!op3->IsAllFalse())
+                {
+                    // Replace op3 with Zero, if not already zero
+                    op3->SetUnusedValue();
+
+                    GenTree** const treeUse  = &(cndSelNode->Op(1));
+                    GenTree*        zeroNode = comp->gtNewZeroConNode(op3->TypeGet());
+
+                    BlockRange().InsertBefore(cndSelNode, zeroNode);
+
+                    cndSelNode->ReplaceOperand(treeUse, zeroNode);
+                }
+                return cndSelNode;
+            }
+            else
+            {
+                // Transform:
+                //
+                // CndSel(AllTrue, op2, op3) to
+                // op2
+                op1->SetUnusedValue();
+                op3->SetUnusedValue();
+                cndSelNode->SetUnusedValue();
+                return op2;
+            }
+        }
+    }
+
+    if (op2->OperIsHWIntrinsic())
+    {
+        // Handle cases where there is a nested ConditionalSelect for
+        // `trueValue`
+        GenTreeHWIntrinsic* trueValNode = op2->AsHWIntrinsic();
+        const HWIntrinsic   intrinEmbMask(trueValNode);
+        if (intrinEmbMask.id == NI_Sve_ConditionalSelect)
+        {
+            GenTree* trueValOp1 = trueValNode->Op(1);
+            GenTree* trueValOp2 = trueValNode->Op(2);
+            GenTree* trueValOp3 = trueValNode->Op(3);
+
+            if (trueValOp1->IsAllTrue() && trueValOp2->isEmbeddedMaskingCompatibleHWIntrinsic())
```

Comment:
The check for `isEmbeddedMaskingCompatibleHWIntrinsic` isn't really necessary. We can just fold any `CndSel(mask, CndSel(true, x, y), z)` into `CndSel(mask, x, z)`.

We would need to ensure that `y` is marked unused, however.

Reply:
done.